### ISSUE TITLE
different probabilities between python and java

### ISSUE THREAD

**Issue by yairdata** [catboost/catboost#1711]
catboost version: 0.25.1
Operating System: linux

i trained a binary classification model, saved it as hbm file , and used the catboost-prediction jar in java for my service that does inference in real-time.
the sanity test with a simple catboost model with few categorical and numeric variables works great.
when i create the actual model i get different probabilities.
i think the issue is the precision.
the numeric variables in java are converted to float since the predict function get string[] and float[].
in python i cast categorical features to 'category' and numeric features to float64. (tried also float32 , didnt work)
i validated that the same input goes in each case to the model , and i see only very minor rounding cases between java & python.
since the probabilities i am looking at are very small , any slight shift can affect the model performance.
how can i sort it out and make the model stable from training to inference time ?

**Comment by kizill** [comment]
Can you prepare a simple reproducer to help us find what's wrong?

**Comment by yairdata** [comment]
training code :
from catboost import CatBoostClassifier
import pandas as pd


X_train = [["a","b",1.1,2.11,1000045552.1233452,0.0023],
          ["a","b",1.1,2.11,100004.3452,0.0023],
          ["c","a",1.221,2.11,5433210004.5423452,0.02023],
          ["a","b",13.1,2.11,100004.3452,0.001123],
          ["a","e",12.1,2.11,10000483621.345222,0.0023],
          ["d","b",1.1,2.11,100004.3452,0.0023],
          ["a","f",1.1,2.11,10055204.004452,0.0023]]

cat_features=[0,1]

y_train=[0,0,1,0,0,1,0]
cols=["one","two","three","four","five","six"]
train = pd.DataFrame(X_train, columns=cols)
label = pd.DataFrame(y_train, columns=["label"])


for col in ["one","two"]:
    train[col]= train[col].astype('category')

model = CatBoostClassifier(
     iterations=100,
     random_seed=63
 )
model.fit(train,label,cat_features)

predictions=model.predict_proba(train)
predictions

outout is:
`array([[0.58540743, 0.41459257],
       [0.58825546, 0.41174454],
       [0.45876698, 0.54123302],
       [0.58150833, 0.41849167],
       [0.58079443, 0.41920557],
       [0.4700186 , 0.5299814 ],
       [0.58918891, 0.41081109]])`


**here is the inference code, testing 3rd line of the training data**
final String modelFileName = "test_catboost.hbm";
        CatBoostModel model = null;
        String modelPath = "H:\\";
        try {
            model = CatBoostModel.loadModel(modelPath+"/"+modelFileName);
        } catch (CatBoostError catBoostError) {
            throw new MlSystemException(
                    String.format("CatBoost model load: {}", modelFileName), catBoostError);
        }
        final String[] categoricalData = new String[]{"c","a"};
        final float[] numericData = new float[]{1.221f,2.11f,5433210004.5423452f,0.02023f};
        try {
            CatBoostPredictions prediction = model.predict(numericData,
                    categoricalData);
            double final_score = sigmoid(prediction.get(0, 0));
            System.out.println(final_score);
         }catch (Exception e){}
**output:**
`0.5299813968301598`
